src/java.desktop/share/native/liblcms/LCMS.c

```diff
@@ -613,50 +613,32 @@ JNIEXPORT void JNICALL Java_sun_java2d_cmm_lcms_LCMS_colorConvert
 /*
  * Class:     sun_java2d_cmm_lcms_LCMS
  * Method:    getProfileID
- * Signature: (Ljava/awt/color/ICC_Profile;)Lsun/java2d/cmm/lcms/LCMSProfile
+ * Signature: (Ljava/awt/color/ICC_Profile;)Lsun/java2d/cmm/lcms/LCMSProfile;
  */
 JNIEXPORT jobject JNICALL Java_sun_java2d_cmm_lcms_LCMS_getProfileID
   (JNIEnv *env, jclass cls, jobject pf)
 {
-    jclass clsLcmsProfile;
-    jobject cmmProfile;
-    jfieldID fid;
-
     if (pf == NULL) {
         return NULL;
     }
-
     jclass pcls = (*env)->GetObjectClass(env, pf);
     if (pcls == NULL) {
         return NULL;
     }
-    jmethodID mid = (*env)->GetMethodID(env, pcls, "activate", "()V");
+    jmethodID mid = (*env)->GetMethodID(env, pcls, "cmmProfile",
+                                        "()Lsun/java2d/cmm/Profile;");
     if (mid == NULL) {
         return NULL;
     }
-    (*env)->CallVoidMethod(env, pf, mid);
+    jobject cmmProfile = (*env)->CallObjectMethod(env, pf, mid);
     if ((*env)->ExceptionOccurred(env)) {
         return NULL;
     }
-
-    fid = (*env)->GetFieldID(env, pcls, "cmmProfile",
-                             "Lsun/java2d/cmm/Profile;");
-    if (fid == NULL) {
-        return NULL;
-    }
-
-    clsLcmsProfile = (*env)->FindClass(env,
-            "sun/java2d/cmm/lcms/LCMSProfile");
-    if (clsLcmsProfile == NULL) {
-        return NULL;
-    }
-
-    cmmProfile = (*env)->GetObjectField (env, pf, fid);
-
-    if (JNU_IsNull(env, cmmProfile)) {
+    jclass lcmsPCls = (*env)->FindClass(env, "sun/java2d/cmm/lcms/LCMSProfile");
+    if (lcmsPCls == NULL) {
         return NULL;
     }
-    if ((*env)->IsInstanceOf(env, cmmProfile, clsLcmsProfile)) {
+    if ((*env)->IsInstanceOf(env, cmmProfile, lcmsPCls)) {
         return cmmProfile;
     }
     return NULL;
```

Comment:
Why do we need to do this from native code? (except easing of access to a private method of a class in another package.)
Will it give some noticeable performance boost if we implement it on java side?

Reply:
Yes, this is the only reason.
I have a todo to check what access will be better, AWTAccessor/methodhandle/reflection vs jni.